### PR TITLE
fixes lda condition for blas functions, fixes bug with beta=0 in addmv slow path

### PR DESCRIPTION
**Pull request by ngimel**
per title. If `beta=0` and slow path was taken, `nan` and `inf` in the result were not masked as is the case with other linear algebra functions. Similarly, since `mv` is implemented as `addmv` with `beta=0`, wrong results were sometimes produced for `mv` slow path.

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -16405,6 +16405,13 @@ def test_addmv(self, device, dtype):
         v = torch.randn(100, device=device).to(dtype)
         self._test_addmm_addmv(torch.addmv, t, m, v, beta=0)
 
+        # Test beta=0, v=nan, 0-strided v
```
Comment (zasdfgbnm) [inline test/test_torch.py:16408]:
```suggestion
        # Test beta=0, t=nan, 0-strided v
```

File: aten/src/ATen/native/BlasKernel.cpp [hunk 2]
```diff
@@ -191,12 +191,15 @@ void gemv(char trans, int64_t m, int64_t n, scalar_t alpha, scalar_t *a, int64_t
...
     }
   } else {
-    if (beta != scalar_t(1)) scal<scalar_t>(m, beta, y, incy);
+    if (beta != scalar_t(1) && beta != scalar_t(0)) scal<scalar_t>(m, beta, y, incy);
 
     for (int64_t j = 0; j < n; j++) {
       scalar_t *column_ = a + lda * j;
       scalar_t z = alpha * x[j * incx];
       for (int64_t i = 0; i < m; i++) {
+        if (j==0 && beta==scalar_t(0)) {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/BlasKernel.cpp:200]: It's better if you leave a comment here.

**Comment by ngimel** [comment]
Tests expanded according to @zasdfgbnm suggestion (thanks!), initialization for broadcasted matrices/vectors in tests fixed to remain within exactly representable numbers and reduce errors

**Review by zasdfgbnm — APPROVED** [review]

File: aten/src/ATen/native/BlasKernel.cpp (hunk 2 above)
Comment (mruberry) [reply aten/src/ATen/native/BlasKernel.cpp:201]: I see your war on spacing continues.

File: test/test_torch.py [hunk 3]
```diff
@@ -16388,22 +16370,36 @@ def test_addmv(self, device, dtype):
...
+            torch.randn(10, device=device).to(dtype),
+            torch.randn(1, device=device).to(dtype).expand(10),
+        ]
+        vs = [
+            torch.randn(100, device=device).to(dtype),
+            torch.ones(1, device=device).to(dtype).expand(100),  # to reduce errors for low precision
+        ]
+        ms = [
+            # 0d
+            torch.ones((), device=device).to(dtype).expand(10, 100),  # to reduce errirs for low precision
```
Comment (mruberry) [inline test/test_torch.py:16383]: "errirs"

File: test/test_torch.py [hunk 4]
```diff
@@ -16388,22 +16388,62 @@ def test_addmv(self, device, dtype):
...
+            # 1d
+            torch.randn((1, 100), device=device).to(dtype).expand(10, 100),
+            # this initialization reduces errors for low precision for broadcasted matrices
```
[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_torch.py:16386]: would you like to elaborate briefly on HOW it does that?

File: test/test_torch.py [hunk 5]
```diff
@@ -16388,22 +16370,36 @@ def test_addmv(self, device, dtype):
...
+            torch.randint(3, (10, 1), dtype=torch.float, device=device).to(dtype).expand(10, 100),
+            torch.randint(3, (1, 10), dtype=torch.float, device=device).to(dtype).expand(100, 10).t(),
+            torch.randn((100, 1), device=device).to(dtype).expand(100, 10).t(),
+            # 2d
+            torch.randn((10, 100), device=device).to(dtype),
+            torch.randn((100, 10), device=device).to(dtype).t(),
+        ]
+        for m in ms:
```
Comment (mruberry) [inline test/test_torch.py:16394]:
```
for m, v, t in product(ms, vs, ts):
```

File: test/test_torch.py [hunk 6]
```diff
@@ -16388,22 +16370,36 @@ def test_addmv(self, device, dtype):
...
+            for v in vs:
+                for t in ts:
+                    self._test_addmm_addmv(torch.addmv, t, m, v)
+        # Test beta=0, t=nan
         t = torch.full((10,), math.nan, device=device).to(dtype)
-        m = torch.randn(10, 100, device=device).to(dtype)
-        v = torch.randn(100, device=device).to(dtype)
-        self._test_addmm_addmv(torch.addmv, t, m, v, beta=0)
+        for m in ms:
```
Comment (mruberry) [inline test/test_torch.py:16400]: Similar opportunity to use product here

**Review by mruberry — APPROVED** [review]
Maybe we can have nice things after all.

**Review by zou3519 — APPROVED** [review]
